### PR TITLE
Check for Architecture during Registry Build

### PR DESCRIPTION
**Pull request by maysunfaisal**
**Please specify the area for this PR**
index/generator

**What does does this PR do / why we need it**:
Checks the architecture list during the registry build and outputs a warning to stdout if the architecture list is empty. Its best practice to mention the architecture when authoring devfiles but its not mandatory. Check https://github.com/devfile/api/issues/244#issuecomment-859732271 for more info

**Which issue(s) this PR fixes**:

Addresses part of comment https://github.com/devfile/api/issues/244#issuecomment-859732271

**PR acceptance criteria**:

- [x] Test (WIP) 

- [ ] Documentation (WIP)

**How to test changes / Special notes to the reviewer**:
build and run 

```
 ./index-generator ../../tests/registry /tmp/index.json
 
the java-maven devfile has no architecture(s) mentioned
the java-quarkus devfile has no architecture(s) mentioned
the nodejs-basic devfile has no architecture(s) mentioned
the code-with-quarkus devfile has no architecture(s) mentioned
the java-springboot-basic devfile has no architecture(s) mentioned
the python-basic devfile has no architecture(s) mentioned
```

Index Example with Arch:

```
  {
    "name": "nodejs",
    "version": "1.0.0",
    "displayName": "NodeJS Runtime",
    "description": "Stack with NodeJS 12",
    "type": "stack",
    "tags": [
      "NodeJS",
      "Express",
      "ubi8"
    ],
    "architectures": [
      "amd64"
    ],
    "projectType": "nodejs",
    "language": "nodejs",
    "links": {
      "self": "devfile-catalog/nodejs:latest"
    },
    "resources": [
      "devfile.yaml"
    ],
    "starterProjects": [
      "nodejs-starter"
    ]
  },
  ```

### CODE REVIEW

File: index/generator/library/library.go [hunk 1]
```diff
@@ -151,7 +164,11 @@ func parseDevfileRegistry(registryDirPath string, force bool) ([]schema.Schema,
 			// Index component validation
 			err := validateIndexComponent(indexComponent, schema.StackDevfileType)
 			if err != nil {
-				return nil, fmt.Errorf("%s index component is not valid: %v", devfileDir.Name(), err)
+				if _, ok := err.(*MissingArchError); !ok {
```
Comment (johnmcollier) [inline index/generator/library/library.go:167]: This section looks fine, but was a little hard to understand at first what the if statement was doing.

To improve clarity, could we do something like the following?
```
if _, ok := err.(*MissingArchError); ok {
    fmt.Printf("the %s devfile has no architecture(s) mentioned\n", indexComponent.Name)
else {
    return nil, fmt.Errorf("%s index component is not valid: %v", devfileDir.Name(), err)
}
```

File: index/generator/library/library.go [hunk 2]
```diff
@@ -214,7 +231,11 @@ func parseExtraDevfileEntries(registryDirPath string, force bool) ([]schema.Sche
 				// Index component validation
 				err := validateIndexComponent(indexComponent, devfileType)
 				if err != nil {
-					return nil, fmt.Errorf("%s index component is not valid: %v", indexComponent.Name, err)
+					if _, ok := err.(*MissingArchError); !ok {
```
Comment (johnmcollier) [inline index/generator/library/library.go:234]: Slight nit, take it or leave it:

This section is duplicated with the above section you added, could we abstract it out to a function?

File: index/generator/library/library.go [hunk 3]
```diff
@@ -151,7 +164,11 @@ func parseDevfileRegistry(registryDirPath string, force bool) ([]schema.Schema,
 			// Index component validation
 			err := validateIndexComponent(indexComponent, schema.StackDevfileType)
 			if err != nil {
-				return nil, fmt.Errorf("%s index component is not valid: %v", devfileDir.Name(), err)
+				if _, ok := err.(*MissingArchError); !ok {
+					return nil, fmt.Errorf("%s index component is not valid: %v", devfileDir.Name(), err)
+				}
+				// log to the console as FYI if the devfile has no architectures
+				fmt.Printf("the %s devfile has no architecture(s) mentioned\n", indexComponent.Name)
```
Comment (johnmcollier) [inline index/generator/library/library.go:171]: Can't we just print out `err.Error()` here?

**Review by johnmcollier — CHANGES_REQUESTED** [review]

File: index/generator/library/library.go (hunk 1 above)
Comment (maysunfaisal) [reply index/generator/library/library.go]: Addressed, made it more readable. However not to the suggested snippet due to warning:
Screen Shot 2021-07-07 at 3 34 58 PM

File: index/generator/library/library.go (hunk 2 above)
Comment (maysunfaisal) [reply index/generator/library/library.go]: gonna leave it, more lines of code(for err checking) if we abstract it

File: index/generator/library/library.go (hunk 3 above)
Comment (maysunfaisal) [reply index/generator/library/library.go]: good point, done

File: index/generator/library/library.go (hunk 2 above)
Comment (johnmcollier) [reply index/generator/library/library.go]: Fair enough

File: index/generator/library/library.go (hunk 1 above)
Comment (johnmcollier) [reply index/generator/library/library.go]: Sounds good to me.

**Review by johnmcollier — APPROVED** [review]

**Pull request merged**